pkg/hook/transformer.go

```diff
@@ -0,0 +1,117 @@
+package hook
+
+import (
+	"go/ast"
+	"go/token"
+	"go/types"
+	"path/filepath"
+	"strconv"
+
+	"golang.org/x/tools/go/ast/astutil"
+)
+
+type Transformer struct {
+	file               *ast.File
+	fileSet            *token.FileSet
+	filePath           string
+	typeInfo           *types.Info
+	transformedImports map[string]string
+}
+
+func NewTransformer(file *ast.File, fileSet *token.FileSet, filePath string, typeInfo *types.Info) *Transformer {
+	return &Transformer{
+		file:               file,
+		fileSet:            fileSet,
+		filePath:           filePath,
+		typeInfo:           typeInfo,
+		transformedImports: make(map[string]string),
+	}
+}
+
+func (t *Transformer) TransformFile() (transformed bool) {
+	astutil.Apply(t.file, nil, func(cursor *astutil.Cursor) bool {
+		callExpr, ok := cursor.Node().(*ast.CallExpr)
+		if !ok {
+			return true
+		}
+
+		selectorExpr, ok := callExpr.Fun.(*ast.SelectorExpr)
+		if !ok {
+			return true
+		}
+
+		receiver, ok := selectorExpr.X.(*ast.Ident)
+		if !ok {
+			return true
+		}
+
+		if pkgName, pkgPath, ok := importedPackage(receiver, t.file); ok {
+			if h := MatchingFunctionHook(selectorExpr.Sel.Name, pkgPath); h != nil {
+				selectorExpr.Sel.Name = h.HookName
+				selectorExpr.X = &ast.Ident{
+					Name: sanitizersPackageName,
+				}
+				callExpr.Args = append([]ast.Expr{FakePC(callExpr, t.filePath, t.fileSet)}, callExpr.Args...)
+
+				astutil.AddNamedImport(t.fileSet, t.file, sanitizersPackageName, sanitizersPackagePath)
+
+				t.transformedImports[pkgPath] = pkgName
+				transformed = true
+			}
+		} else if receiverType := t.typeInfo.Types[receiver]; receiverType.Addressable() {
+			if h := MatchingMethodHook(selectorExpr.Sel.Name, receiverType.Type.String()); h != nil {
+				selectorExpr.Sel.Name = h.HookName
+				selectorExpr.X = &ast.Ident{
+					Name: sanitizersPackageName,
+				}
+				callExpr.Args = append([]ast.Expr{FakePC(callExpr, t.filePath, t.fileSet), receiver}, callExpr.Args...)
+				astutil.AddNamedImport(t.fileSet, t.file, sanitizersPackageName, sanitizersPackagePath)
+				transformed = true
+			}
+		}
+
+		return true
+	})
+
+	if transformed {
+		t.removeUnusedImport()
+	}
+	return
+}
+
+func (t *Transformer) removeUnusedImport() {
+	for path, name := range t.transformedImports {
+		if !astutil.UsesImport(t.file, path) {
+			astutil.DeleteNamedImport(t.fileSet, t.file, name, path)
+		}
+	}
+}
+
+func importedPackage(ident *ast.Ident, file *ast.File) (name string, path string, found bool) {
```

Comment:
The idea was to add named returns for cases where it is confusing as in this example where we have two strings and a boolean. Could you mention which cases do you find confusing so that they can be adjusted?